pkg/metricsapiservice/register.go

```diff
@@ -0,0 +1,134 @@
+package metricsapiservice
+
+import (
+	"context"
+	"math"
+	"time"
+
+	vclustercontext "github.com/loft-sh/vcluster/cmd/vcluster/context"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/rest"
+	"k8s.io/klog"
+	apiregistrationv1 "k8s.io/kube-aggregator/pkg/apis/apiregistration/v1"
+	"k8s.io/metrics/pkg/apis/metrics"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	apiregclientv1 "k8s.io/kube-aggregator/pkg/client/clientset_generated/clientset/typed/apiregistration/v1"
+)
+
+const (
+	MetricsVersion    = "v1beta1"
+	MetricsAPIService = MetricsVersion + "." + metrics.GroupName // "v1beta1.metrics.k8s.io"
+
+	KubernetesSvc = "kubernetes"
+)
+
+func checkExistingAPIService(ctx context.Context, kubeAggClient *apiregclientv1.ApiregistrationV1Client) bool {
+	var exists bool
+
+	_ = applyOperation(ctx, func() (bool, error) {
+		_, err := kubeAggClient.APIServices().Get(ctx, MetricsAPIService, v1.GetOptions{})
+		if err != nil {
+			if kerrors.IsNotFound(err) {
+				return true, nil
+			}
+
+			return false, err
+		}
+
+		exists = true
+		return true, nil
+	})
+
+	return exists
+}
+
+func applyOperation(ctx context.Context, operationFunc wait.ConditionFunc) error {
+	return wait.ExponentialBackoffWithContext(ctx, wait.Backoff{
+		Duration: time.Second,
+		Factor:   1.5,
+		Cap:      time.Minute,
+		Steps:    math.MaxInt32,
+	}, operationFunc)
+}
+
+func deleteOperation(ctx context.Context, kubeAggClient *apiregclientv1.ApiregistrationV1Client) wait.ConditionFunc {
+	return func() (bool, error) {
+		err := kubeAggClient.APIServices().Delete(ctx, MetricsAPIService, v1.DeleteOptions{})
+		if err != nil {
+			if kerrors.IsNotFound(err) {
+				return true, nil
+			}
+
+			return false, err
+		}
+
+		return true, nil
+	}
+}
+
+func createOperation(ctx context.Context, kubeAggClient *apiregclientv1.ApiregistrationV1Client, client client.Client) wait.ConditionFunc {
+	return func() (bool, error) {
+		spec := apiregistrationv1.APIServiceSpec{
+			Group:                metrics.GroupName,
+			GroupPriorityMinimum: 100,
+			Version:              MetricsVersion,
+			VersionPriority:      100,
+		}
+
+		apiService := &apiregistrationv1.APIService{
+			ObjectMeta: v1.ObjectMeta{
+				Name: MetricsAPIService,
+			},
+		}
+
+		_, err := controllerutil.CreateOrUpdate(ctx, client, apiService, func() error {
+			apiService.Spec = spec
+			return nil
+		})
+		if err != nil {
+			if kerrors.IsAlreadyExists(err) {
+				return true, nil
+			}
```

Comment:
nit-pick, don't even bother changing this if there is nothing else to change :D 
I don't think this type of error can happen after the change to CreateOrUpdate :)